### ISSUE TITLE
Remapping <c-e>, <c-y>, <cr> does not seem to be necessary

### ISSUE THREAD

**Issue by 720720** [lifepillar/vim-mucomplete#66]
- [x] Output of `setl completeopt`: completeopt=menuone,preview,noinsert
- [x] MUcomplete settings in my `.vimrc`, if any: let mucomplete#enable_auto_at_startup = 1

For automatic completion, you also need to put these in your `vimrc`:
```
inoremap <expr> <c-e> mucomplete#popup_exit("\<c-e>")
inoremap <expr> <c-y> mucomplete#popup_exit("\<c-y>")
inoremap <expr>  <cr> mucomplete#popup_exit("\<cr>")
```
Could you please explain these mappings. Vim 8.0 shows the same behavior with or without them.

**Comment by lifepillar** [comment]
The purpose of the mappings is to allow µcomplete to detect when the pop-up menu is closed by the user with `<c-e>`, `<c-y>` or `<cr>`, in which case auto-completion must not kick in again (otherwise, the user would not be able to dismiss the pop-up menu). What `mucomplete#popup_exit()` does is just record in a script variable whether the pop-up menu is visible or not at the time the key is pressed.

Such mappings were introduced before I changed the auto-completion mechanism to use `InsertCharPre` and `TextChangedI` together. Since `InsertCharPre` is not triggered when `<c-e>`, `<c-y>` or `<cr>` is pressed, It is entirely possible that they are not necessary any more, but I have to verify that, because there might be some edge cases.

**Comment by lifepillar** [comment]
It turns out that the mappings are necessary. To see why, make sure autocompletion is on and that `<c-e>` is unmapped in Insert mode:

```vim
:MUcompleteAutoOn
:iunmap <c-e>
```

then type the following (I assume that local keyword completion is in your completion chain):

```
foobar foo<c-e>
```

Note that the pop-up menu is not closed. You need to type `<c-e>` twice.

The following is a detailed technical explanation of what happens under the hood, in case you are interested.

MUcomplete hooks into both `InsertCharPre` and `TextChangedI`. Note that `InsertCharPre` is fired *before* a character is inserted and it is fired even when the pop-up menu is visible, but it is not fired on `<c-e>`, `<c-y>` or `<cr>`. Instead, `TextChangedI` is fired *after* any text change, but it is not fired while the pop-up menu is visible. MUcomplete takes advantage of the two events as follows:

- on `InsertCharPre`, µcomplete just sets `s:insertcharpre` to `1`;
- on `TextChangedI` µcomplete invokes `s:act_on_textchanged()`. This function checks the value of `s:insertcharpre`: if it finds that it is `1`, it sets it back to `0` and tries to auto-complete the current text (if `s:insertcharpre` is `0` when `s:act_on_textchanged()` is called, then the function does nothing).

So, this is how the example above is processed:

1. up to `foobar f`, nothing relevant happens.
2. When typing `o`, `InsertCharPre` is fired and `s:insertcharpre` is set to `1`, then `TextChangedI` is fired and `s:act_on_textchanged()` is invoked, which resets `s:insertcharpre` and triggers keyword completion. The pop-up menu becomes visible.
3. When the user types the second `o`, `InsertCharPre` is fired again and `s:insertcharpre` is set to `1`. Nothing else happens: since the pop-up menu is visible, `TextChangedI` is not fired.
4. Now the user presses `<c-e>`. This dismisses the pop-up menu and triggers `TextChangedI`. Since `s:insertcharpre` is currently `1`, µcomplete resets `s:insertcharpre` and triggers keyword completion, causing the pop-up menu to appear again.

By mapping `<c-e>` (and `<c-y>` and `<cr>`), µcomplete is able to skip auto-completion at step 4.

**Comment by lifepillar** [comment]
In fact, by not setting `s:insertcharpre` when the pop-up menu is visible should make the mappings redundant.

**Comment by lifepillar** [comment]
>In fact, not setting s:insertcharpre when the pop-up menu is visible should make the mappings redundant.

That's not the case, actually. MUcomplete must set `s:insertcharpre` even when the pop-up menu is visible to be able to trigger auto-completion when the pop-up menu is closed by a key different from `<c-e>`, `<c-y>` or `<cr>`. To see that, create a buffer with the text `locaal`, turn on auto-completion, and type:

```
/usr/locaal
```

This text first triggers path completion, so that after `/usr/loca` the pop-up menu shows a list of paths. When you press the second `a`, though, the following happens:

1. `InsertCharPre` is fired and `s:insertcharpre` is set to `1`.
2. Since no path matches, the menu is closed.
3. `TextChangedI` is fired. Since `s:insertcharpre` is `1`, µcomplete triggers auto-completion again and offers keyword completion.

If `s:insertcharpre` were not set when the pop-up menu is visible (step 1), at step 3 µcomplete would not be able to immediately offer keyword completion (`locaal`).

To summarize, the thin wrappers around `<c-e>`, `<c-y>` and `<cr>` are needed to determine how the pop-up menu is closed. If it is closed by inserting a visible key, then auto-completion must be triggered again; otherwise, it must not.
